errors.go

```diff
@@ -215,6 +235,40 @@ func (errorType ErrorType) UserErrorf(msg string, args ...interface{}) error {
 	}
 }
 
+// AddDetails adds a details element to an error
+func (errorType ErrorType) AddDetails(err error, details interface{}) error {
+	if err == nil {
+		return errorType.details(details)
+	}
+
+	if details == nil {
+		return err
+	}
+
+	c := new(customError)
+	c.error = errors.WithStack(err)
+	c.userMessage = GetUserMessage(err)
+
+	c.details = append(GetDetails(err), details)
+
+	if errorType != NoType {
+		c.errorType = errorType
+	} else {
+		c.errorType = GetType(err)
+	}
+
+	return c
+}
+
+// details creates a new error with arbitrary details
+func (errorType ErrorType) details(details interface{}) error {
+	return &customError{
```

Comment:
We should add 
``` 
if details == nil {
  return err
}
```

otherwise,  if someone is calling `AddDetails(nil,nil)`
they would get `Details: [{}]` and we don't want that

Reply:
good point. adding it now

Reply:
Added it at the begining of `AddDetails`, as `func details()` is only useful when `err == nil`